File: src/main/java/innexgo/hours/LocationService.java

/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class LocationService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public Location getByLocationId(long locationId) {
    String sql =
        "SELECT * FROM location WHERE location_id=?";
    RowMapper<Location> rowMapper = new LocationRowMapper();
    Location location = jdbcTemplate.queryForObject(sql, rowMapper, locationId);
    return location;
  }

  public long nextId() {
    String sql = "SELECT max(location_id) FROM location";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }

  public void add(Location location) {
    location.locationId = nextId();
    // Add location
    String sql = "INSERT INTO location values (?,?,?,?,?,?,?)";
    jdbcTemplate.update(
        sql,
        location.locationId,
        location.creationTime,
        location.creatorUserId,
        location.schoolId,
        location.name,
        location.description,
        location.valid);
  }

  public boolean existsByLocationId(long locationId) {
    String sql = "SELECT count(*) FROM location WHERE location_id=?";
    int count = jdbcTemplate.queryForObject(sql, Integer.class, locationId);
    return count != 0;
  }

 public List<Location> query( //
     Long locationId, //
     Long creationTime, //
     Long minCreationTime, //
     Long maxCreationTime, //
     Long creatorUserId, //
     Long schoolId, //
     String name, //
     String partialName, //
     String description, //
     Boolean valid, //
     long offset, //
     long count) //
 {

    String sql =
      "SELECT l.* FROM location l"
        + " WHERE 1=1 "
        + (locationId      == null ? "" : " AND l.location_id = " + locationId)
        + (creationTime    == null ? "" : " AND l.creation_time = " + creationTime)
        + (minCreationTime == null ? "" : " AND l.creation_time > " + minCreationTime)
        + (maxCreationTime == null ? "" : " AND l.creation_time < " + maxCreationTime)
        + (creatorUserId   == null ? "" : " AND l.creator_user_id = " + creatorUserId)
        + (schoolId        == null ? "" : " AND l.school_id = " + schoolId)
        + (name            == null ? "" : " AND l.name = " + Utils.escape(name))
        + (partialName     == null ? "" : " AND l.name LIKE " + Utils.escape("%"+partialName+"%"))
        + (description     == null ? "" : " AND l.description = " + Utils.escape(description))
        + (valid           == null ? "" : " AND l.valid = " + valid)
        + (" ORDER BY l.location_id")
        + (" LIMIT " + offset + ", " + count)
        + ";";

    RowMapper<Location> rowMapper = new LocationRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper);
  }
}

File: src/location_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use tokio_postgres::GenericClient;
use innexgo_hours_api::request;

impl From<tokio_postgres::row::Row> for Location {
  // select * from location order only, otherwise it will fail
  fn from(row: tokio_postgres::row::Row) -> Location {
    Location {
      location_id: row.get("location_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      school_id: row.get("school_id"),
    }
  }
}

pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  school_id: i64,
) -> Result<Location, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let location_id = con
    .query_one(
      "INSERT INTO
       location_t(
           creation_time,
           creator_user_id,
           school_id
       )
       VALUES($1, $2, $3)
       RETURNING location_id
      ",
      &[&creation_time, &creator_user_id, &school_id],
    ).await?
    .get(0);

  // return location
  Ok(Location {
    location_id,
    creation_time,
    creator_user_id,
    school_id
  })
}

pub async fn get_by_location_id(
  con: &mut impl GenericClient,
  location_id: i64,
) -> Result<Option<Location>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM location_t WHERE location_id=$1",
      &[&location_id],
    ).await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: request::LocationViewProps,
) -> Result<Vec<Location>, tokio_postgres::Error> {
  let results = con
    .query(
      "
        SELECT lc.* FROM location_t lc WHERE 1 = 1
        AND ($1::bigint[] IS NULL OR lc.location_id = ANY($1))
        AND ($2::bigint   IS NULL OR lc.creation_time >= $2)
        AND ($3::bigint   IS NULL OR lc.creation_time <= $3)
        AND ($4::bigint[] IS NULL OR lc.creator_user_id = ANY($4))
        AND ($5::bigint[] IS NULL OR lc.school_id = ANY($5))
        ORDER BY lc.location_id
      ",
      &[
        &props.location_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.school_id,
      ],
    ).await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
